hyperspy/_signals/wave_image.py

```diff
@@ -0,0 +1,165 @@
+#
+# This file is part of  HyperSpy.
+#
+#  HyperSpy is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+#  HyperSpy is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with  HyperSpy.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import numpy as np
+
+from skimage.restoration import unwrap_phase as unwrap
+
+from hyperspy._signals.image import Image
+
+
+class WaveImage(Image):
+    """Image subclass for complex electron wave data (e.g. reconstructed from holograms)."""
+
+    _signal_type = 'WAVE'
+
+    @property
+    def phase(self):
+        """Get/set the phase of the data. Returns an :class:`~hyperspy._signals.Image`."""
+        phase = self._deepcopy_with_new_data(np.angle(self.data))
+        phase.set_signal_type('')  # Result is a normal Image!
+        return phase
+
+    @phase.setter
+    def phase(self, phase):
+        if isinstance(phase, Image):
+            phase = phase.data
+        self.data = self.amplitude.data * np.exp(1j * phase)
+
+    @property
+    def amplitude(self):
+        """Get/set the amplitude of the data. Returns an :class:`~hyperspy._signals.Image`."""
+        amplitude = self._deepcopy_with_new_data(np.abs(self.data))
+        amplitude.set_signal_type('')  # Result is a normal Image!
+        return amplitude
+
+    @amplitude.setter
+    def amplitude(self, amplitude):
+        if isinstance(amplitude, Image):
+            amplitude = amplitude.data
+        self.data = amplitude * np.exp(1j * self.phase.data)
+
+    @property
+    def real(self):
+        """Get/set the real part of the data. Returns an :class:`~hyperspy._signals.Image`."""
+        real = self._deepcopy_with_new_data(np.real(self.data))
+        real.set_signal_type('')  # Result is a normal Image!
+        return real
+
+    @real.setter
+    def real(self, real):
+        if isinstance(real, Image):
+            real = real.data
+        self.data = real + 1j * self.imag.data
+
+    @property
+    def imag(self):
+        """Get/set the imaginary part of the data. Returns an :class:`~hyperspy._signals.Image`."""
+        imag = self._deepcopy_with_new_data(np.imag(self.data))
+        imag.set_signal_type('')  # Result is a normal Image!
+        return imag
+
+    @imag.setter
+    def imag(self, imag):
+        if isinstance(imag, Image):
+            imag = imag.data
+        self.data = self.real.data + 1j * imag
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Make sure data is complex:
+        self.change_dtype(complex)

```

Comment:
Why changing the dtype in the constructor? This comes with dangers, as there is more than one complex dtype.